### PR TITLE
Expose kms timeout value via encryption config.

### PR DESCRIPTION
**Pull request by immutableT**
**What type of PR is this?**
>
 /kind feature
**What this PR does / why we need it**:
Expose, via EncryptionConfiguration, the timeout value that kube-apiserver will use when awaiting responses from kms-plugin.

This is a followup to 68585, were a timeout was added to calls from kube-apiserver to kms-plugin. This PR exposes the value of the timeout via EncryptionConfiguration. Providing the appropriate value of the timeout should be left to cluster managers who are familiar with the environment in which the cluster and the plugin operate. For example, KMS rooted in in a hardware device like (HSM/TPM) may have fairly long delays (seconds), while in memory KMS implementations would have timeouts in the range of milliseconds.
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kube-apiserver: When configuring integration with external KMS Providers, users  can supply timeout value (i.e. how long should kube-apiserver wait before giving up on a call to KMS).  (@immutableT )
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/apis/config/types.go [hunk 1]
```diff
@@ -123,4 +123,6 @@ type KMSConfiguration struct {
 	CacheSize int32
 	// endpoint is the gRPC server listening address, for example "unix:///var/run/kms-provider.sock".
 	Endpoint string
+	// Timeout for gRPC calls to kms-plugin (ex. 5s).
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: document the default if unspecified

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go [hunk 2]
```diff
@@ -84,4 +84,6 @@ type KMSConfiguration struct {
 	CacheSize int32 `json:"cachesize,omitempty"`
 	// endpoint is the gRPC server listening address, for example "unix:///var/run/kms-provider.sock".
 	Endpoint string `json:"endpoint"`
+	// Timeout for gRPC calls to kms-plugin (ex. 5s).
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go]: document the default if unspecified

File: staging/src/k8s.io/apiserver/pkg/apis/config/types.go [hunk 3]
```diff
@@ -123,4 +123,6 @@ type KMSConfiguration struct {
...
+	Timeout metav1.Duration
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/config/types.go]: should be `*metav1.Duration` so we can tell the difference between "0" and unspecified

File: staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go [hunk 4]
```diff
@@ -84,4 +84,6 @@ type KMSConfiguration struct {
...
+	Timeout metav1.Duration `json:"timeout"`
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go]: should be `+optional` and `omitempty`, right?

**Comment by immutableT** [comment]
@liggitt PTAL.

**Comment by liggitt** [comment]
The failure is legitimate. The generated files need to be regenerated and checked in

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 5]
```diff
@@ -173,8 +173,13 @@ func GetPrefixTransformers(config *apiserverconfig.ResourceConfiguration) ([]val
 				return nil, fmt.Errorf("remote KMS provider can't use empty string as endpoint")
 			}
 
+			timeout := kmsPluginConnectionTimeout
+			if provider.KMS.Timeout != nil {
+				timeout = provider.KMS.Timeout.Duration
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: if specified, verify the duration is greater than zero

**Comment by liggitt** [comment]
one comment on validation, then squash, lgtm otherwise

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 6]
```diff
@@ -173,8 +173,16 @@ func GetPrefixTransformers(config *apiserverconfig.ResourceConfiguration) ([]val
 				return nil, fmt.Errorf("remote KMS provider can't use empty string as endpoint")
 			}
 
+			timeout := kmsPluginConnectionTimeout
+			if provider.KMS.Timeout != nil {
+				if provider.KMS.Timeout.Duration < 0 {
```
[the remark this answers is not in the document]
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: @liggitt good catch - zero seems to make gRPC calls to kms-plugin fail no matter what - even when plugin is up and running before kube-apiserver makes its first call.
Based on that, changed the validation logic to prohibit zero.
Added unit test to cover timeout test cases; let me know if I should put them on a separate PR.

[the remark this answers is not in the document]
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: @liggitt PTAL.

**Comment by liggitt** [comment]
/lgtm
/approve

please detail the config field in the release note and open a doc PR to update the encryption docs for 1.14

**Comment by immutableT** [comment]
Documentation PR 12158 submitted.
https://github.com/kubernetes/website/pull/12158

**Comment by rtheis** [comment]
@liggitt is this something that could be cherry-picked to 1.13?

**Pull request merged**